src/main/scala/com/microsoft/ml/spark/lightgbm/dataset/DatasetAggregator.scala

```diff
@@ -0,0 +1,488 @@
+  FloatChunkedArray, FloatSwigArray, IntChunkedArray, IntSwigArray}
+import org.apache.spark.sql.Row
+
+import scala.collection.mutable.ListBuffer
+
+object ChunkedArrayUtils {
+  def copyChunkedArray[T](chunkedArray: ChunkedArray[T],
+                          mainArray: BaseSwigArray[T],
+                          threadRowStartIndex: Long,
+                          chunkSize: Long): Unit = {
+    // Copy in parallel on each thread
+    // First copy full chunks
+    val chunkCount = chunkedArray.getChunksCount() - 1
+    for (chunk <- 0L until chunkCount) {
+      for (in_chunk_idx <- 0L until chunkSize) {
+        mainArray.setItem(threadRowStartIndex + chunk * chunkSize + in_chunk_idx,
+          chunkedArray.getItem(chunk, in_chunk_idx, (-1).asInstanceOf[T]))
+      }
+    }
+    // Next copy filled values from last chunk only
+    val lastChunkCount = chunkedArray.getLastChunkAddCount()
+    for (last_chunk_idx <- 0L until lastChunkCount) {
+      mainArray.setItem(threadRowStartIndex + chunkCount * chunkSize + last_chunk_idx,
+        chunkedArray.getItem(chunkCount, last_chunk_idx, (-1).asInstanceOf[T]))
+    }
+  }
+}
+
+trait AggregatedColumns
+
+class BaseChunkedColumns(columnParams: ColumnParams,
+                         chunkSize: Int) extends AggregatedColumns {
+  val labelsChunkedArray = new FloatChunkedArray(chunkSize)
+  val weightChunkedArrayOpt = columnParams.weightColumn.map {
+    _ => new FloatChunkedArray(chunkSize)
+  }
+  val initScoreChunkedArrayOpt = columnParams.initScoreColumn.map {
+    _ => new DoubleChunkedArray(chunkSize)
+  }
+  val groupColumnValues: ListBuffer[Row] = new ListBuffer[Row]()
+
+  var rowCount = 0
+
+  def release(): Unit = {
+    // Clear memory
+    labelsChunkedArray.delete()
+    weightChunkedArrayOpt.foreach(_.delete())
+    initScoreChunkedArrayOpt.foreach(_.delete())
+  }
+}
+
+class SparseChunkedColumns(columnParams: ColumnParams,
+                           chunkSize: Int) extends BaseChunkedColumns(columnParams, chunkSize) {
+  var indexesChunkedArray = new IntChunkedArray(chunkSize)
+  var valuesChunkedArray = new DoubleChunkedArray(chunkSize)
+  var indptrChunkedArray = new IntChunkedArray(chunkSize)
+
+  var numCols = 0
+
+  def setNumCols(numCols: Int): Unit = {
+    this.numCols = numCols
+  }
+
+  override def release(): Unit = {
+    // Clear memory
+    super.release()
+    indexesChunkedArray.delete()
+    valuesChunkedArray.delete()
+    indptrChunkedArray.delete()
+  }
+}
+
+class DenseChunkedColumns(columnParams: ColumnParams,
+                          chunkSize: Int,
+                          val numCols: Int) extends BaseChunkedColumns(columnParams, chunkSize) {
+  var featuresChunkedArray = new DoubleChunkedArray(numCols * chunkSize)
+
+  override def release(): Unit = {
+    // Clear memory
+    super.release()
+    featuresChunkedArray.delete()
+  }
+}
+
+class BaseAggregatedColumns extends AggregatedColumns {
```

Comment:
Yes, I'm not sure how to do two passes reliably per partition where the rows remain the same, there's no guarantee in spark as far as I can tell.  Also even doing a count per partition operation is quite slow.  Allocating the chunked arrays seems best here.  There is the other option I wanted to add which is to create the initial lightgbm dataset from a sample and push rows via streaming which we briefly discussed.

Reply:
Once you have a dataset or an RDD or whatever its partitioning structure does not change. Even the situations in which a transformation of a dataset into another dataset will result in a different partitioning structure are fairly limited. I'm pretty sure this is safe.

Reply:
yes I would be very careful here, if I call mapPartitions twice (one to do the count and the other to train) I don't think there is any guarantee the rows will be the same in each partition, especially if dynamic allocation is enabled etc, and I'm not even sure how I would store them so on the next call I could use the row count.  I can't iterate over the iterator twice in scala unless I turn it into an array.  Note this is by design in spark because the iterator is a stream of data (so you can read from disk or any streaming data source without holding everything in memory) so there is no easy way to get the count without actually iterating over all rows, and there is no way to iterate multiple times in a single mapPartitions call.